### ISSUE TITLE
cuda stream issue when using NVCUVID

### ISSUE THREAD

**Issue by robosina** [opencv/opencv#24009]
### System Information

OpenCV version: 4.8.0
Operating System / Platform: Windows 10
Compiler & compiler version: Visual Studio 2017 

Further information: opencv is built with visual studio generator not ninja

### Detailed description

I'm facing an issue when I initiate multiple threads, each using it's own `cv::Ptr<cv::cudacodec::VideoReader>` object, to read from an RTSP camera. I don't utilize the stream object in the function `nextFrame`:

    CV_WRAP virtual bool nextFrame(CV_OUT GpuMat& frame, Stream &stream = Stream::Null()) = 0;
 
which I assume should allow CPU-side synchronization. The idea is to read a `GpuMat` from this function, store it in a `std::dequeue`, and then retrieve a frame from the `std::dequeue` when needed(for processing)

However, problems arise when the `GpuMat` undergoes processing, such as car detection in a deep learning model in a optimized `TensorRT` model which the decoding part I wrote in `CUDA`. I encounter errors indicating `operation not permitted when stream is capturing`. I believe I should be able to fetch frames and process them asynchronously without such issues.

      c:\sources\opencv_4.8.0\opencv_contrib\modules\cudev\include\opencv2\cudev\grid\detail/transform.hpp:267: error: (-217:Gpu API call) operation not permitted when stream is capturing in function 'cv::cudev::grid_transform_detail::TransformDispatcher<false, Policy>::call

### Steps to reproduce

It's challenging to provide sample code, but I'll do my best to deliver and maybe that is some mistakes in my code.

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by cudawarped** [comment]
> It's challenging to provide sample code, but I'll do my best to deliver and maybe that is some mistakes in my code.

I have also observed some strange behaviour when streaming from RTSP in multiple threads with CUDA processing of the resulting `GpuMat`'s but it can take days to appear.  If you could construct a reproducer which fails in a much shorter time let me know as this will help to track down the cause.

**Comment by robosina** [comment]
@cudawarped , Thanks for the prompt answering, For me, it takes around 30 minutes to spot the errors, which unfortunately can't be trapped using `try-catch`, leading to application crashes. As my code is part of proprietary software, I'll create a simulation for you to test.

**Comment by cudawarped** [comment]
> For me, it takes around 30 minutes to spot the errors, which unfortunately can't be trapped using `try-catch`, leading to application crashes

I'm supprised you can't catch the exception?  From memory my problem was that once I caught the exception, reseting the device didn't work so it was of no use.

> As my code is part of proprietary software, I'll create a simulation for you to test. 

Great, if you can reproduce it with CUDA only (i.e. without any inference) that would be better.

> If I don't use the gpuMat in any processing I will not see any error.

Interesting so if you pass the memory around using pointers to your CUDA routines everything is fine.  Maybe the memory inside the `GpuMat` is being destroyed/corrupted automatically somewhere.

**Comment by robosina** [comment]
> Great, if you can reproduce it with CUDA only (i.e. without any inference) that would be better.

Yeah, I will try my best to minimize dependency to able to test it, but it would take some time.

> Interesting so if you pass the memory around using pointers to your CUDA routines everything is fine. Maybe the memory inside the GpuMat is being destroyed/corrupted automatically somewhere.

I'm using a producer/consumer pattern where the consumer receives a GPUMat and sends it to the processor for frame processing. When I disable the processing part on the consumer side, the error doesn't occur. However, the error appears when I use the processor, even though I've cloned the entire GPUMat. Theoretically, there shouldn't be any interference between the producer and consumer when I've cloned the cpuMat? right?, so this error is quite unexpected.

**Comment by cudawarped** [comment]
> Theoretically, there shouldn't be any interference between the producer and consumer when I've cloned the cpuMat? right?,

Theoretically yes, but then everything is happening on the same device.

It sounds like the same issue I was having.  I only seemed to observed errors when I had CUDA processing on my consumer thread otherwise it was fine.  I had a feeling that the issue might be caused by the amount of GPU load from the CUDA side possibly interfering with the timing on the decoding but I could never reproduce the error consistenly enough to verify.  Ideally I wanted to reproduce in a few seconds so I could use Nsight Systems to profile the whole process.

It could also be caused by the WDDM driver but I had the Windows [TDR](https://learn.microsoft.com/en-us/windows-hardware/drivers/display/timeout-detection-and-recovery) disabled while testing.

**Comment by cudawarped** [comment]
> In the above diagram If processing enabled I will see this error, but the exception has been seen in the producer side which is normal to me.

You need to check all CUDA routines for errors because kernel launches are asychronous, therefore you may genereate the error on the consumer side but the next call to the consumer side will return the error from the device.

What kind of synchronization do you have between the producer and consumer?

**Comment by cudawarped** [comment]
> > It could also be caused by the WDDM driver but I had the Windows [TDR](https://learn.microsoft.com/en-us/windows-hardware/drivers/display/timeout-detection-and-recovery) disabled while testing.
> 
> You mentioned the possibility of the WDDM driver causing this issue and that you had TDR disabled while testing. Did disabling TDR offer any improvement? In my case, it actually led to increased system instability.

With TDR enabled the error took longer to appear but I saw a lot of Intel display driver recovery messages in Windows Event Viewer.
```
Event 4101 -Display driver igfx stopped responding and has successfully recovered.
```

Once it was disabled I only saw nvlddmkm restart errors
```
The description for Event ID 0 from source nvlddmkm cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.

If the event originated on another computer, the display information had to be saved with the event.

The following information was included with the event: 

\Device\Video5
Error occurred on GPUID: 100

The message resource is present but the message was not found in the message table
```

I also 
1) removed both the Intel and Nvidia driver completely with [DDU uninstaller](https://www.guru3d.com/files-details/display-driver-uninstaller-download.html) and only reinstalled the Nvidia driver incase the Intel driver was causing the issue, and
2) disabled any antivirus software in case this was causing a delay resulting in a timeout.

I found forcing all the kernel launches to be sychronous with respect to the host by setting the environmental variable [CUDA_LAUNCH_BLOCKING](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=CUDA_LAUNCH_BLOCKING#concurrent-execution-between-host-and-device)=1 appeard to remove the error, but this could simply have delayed it happening.

> I concur that both my custom kernels and OpenCV kernels launch asynchronously. Is this what you were referring to? Could this possibly be causing the issue? if yes do you have any idea to fix it?

No.  I was just pointing out that the function returning the error (e.g. a call to cudaGetLastError or cudaDeviceSynchronize) could be in a different host thread to the one which launched the kernel (or other runtime api function) which caused the error.  This is by design and not an issue just something to be aware of.

**Comment by cudawarped** [comment]
Are you using QT when you see the error?

**Comment by robosina** [comment]
@cudawarped , Thanks for detailed explanation about the problem, 

> [CUDA_LAUNCH_BLOCKING](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=CUDA_LAUNCH_BLOCKING#concurrent-execution-between-host-and-device)=1

That's a great suggestion, I've used it in the past but it slipped my mind. Thanks for reminding me. I'll give it a try and report back with the results, although I suspect it may lead to some performance loss.

> Are you using QT when you see the error?

Regarding the UI, it's designed in C#, with the backend in C++. I do have some simple, non-multithreaded Qt apps where I've not observed errors(since it was not multi-thread)

**Comment by cudawarped** [comment]
> Regarding the UI, it's designed in C#, with the backend in C++. I do have some simple, non-multithreaded Qt apps where I've not observed errors(since it was not multi-thread)

OK that rules out QT.

**Comment by robosina** [comment]
@cudawarped, Your reminder made me think about using CUDA_LAUNCH_BLOCKING=1 even in production mode. I know it's mainly for debugging and not great for the final product, but it doesn't slow down my app too much, just about 3-5%.

Also, I found that I'm using a general (void*) pointer to send the CUDA image in one of my functions. Even though it's not the best practice in C++, I did this to keep the structure of the class. Getting rid of this pointer helped to cut down on some errors.

Lastly, my program can now run for hours without any issues. I also want to point out, there's no problem with dangling references here when using void*. I'm investigating about this matter.

**Comment by cudawarped** [comment]
> It's possible that transferring `cv::cuda::GpuMat` using a `void*` pointer doesn't align with the `RAII` (Resource Acquisition Is Initialization) pattern, and as a result, the reference count might not increase as expected. However, I am not entirely certain about this hypothesis.

I think this could be your issue depending on whether the `GpuMat` is destroyed or not in your producer thread.

If you are passing the `GpuMat::data` pointer from the producer to the consumer then the destructor for the `GpuMat` object can be called in the producer thread before or in between CUDA functions which use that memory in the consumer thread. From my testing in Windows this doesn't always result in a CUDA runtime error, so it is possible that this memory could get re-used somewhere before or during its use in your consumer thread.

The example below shouldn't throw any errors even though the memory has been deleted before calling `cuda::threshold`.  In your scenario `srcFromPtr` would be used in the consumer thread but the memory it points can be altered at any time by `newAlloc`.


```
const Size sz(512, 512);
const int type = CV_8U;
GpuMat dst(sz, type), srcFromPtr;

{
    // alloc new memory and create srcFromPtr using raw memory address
    GpuMat src(sz, type);
    src.setTo(101);
    srcFromPtr = GpuMat(sz, type, src.data, src.step);
} // call src destructor cudaFree(src.data)

// try CUDA function with deleted memory
cuda::threshold(srcFromPtr, dst, 100, 255, THRESH_BINARY);

// no errors above, double check be synchronizing the device
if (cudaDeviceSynchronize() != cudaSuccess) return -1;

// still not errors try to check if the memory is re-allocated by requesting a new GpuMat of the same size
GpuMat newAlloc(sz, type);
if(srcFromPtr.data == newAlloc.data)
    cout << "Memory Deleted: " << (void*)srcFromPtr.data << " == " << (void*)newAlloc.data << endl;
}
```

**Comment by robosina** [comment]
> I think this could be your issue depending on whether the GpuMat is destroyed or not in your producer thread.

Absolutely, your logic makes sense. If the reference count doesn't drop to 0, we indeed can reuse the same memory. However, upon further exploration, I discovered the issue lies with the ONNX C++ runtime. I have a model that doesn't yield correct results with TensorRT, so I've switched to ONNX CUDA for now. The ONNX inference is designed to operate as a class, with each thread requiring its own instance. I noticed that when I instantiate two instances of this class, the program begins to display the previously observed errors. To rectify this, I transformed it into a singleton class, which remarkably eliminated the errors. Therefore, I'm confident that the root cause is linked to ONNX. Further investigation is needed to understand the exact reason behind this issue.

**Comment by halil-karabacak** [comment]
Could this be related to cv::cuda::Stream objects are not thread safe? https://github.com/opencv/opencv/pull/6903/files
